mlflow/telemetry/client.py

```diff
@@ -1,33 +1,251 @@
+        self._lock = threading.RLock()
+        self._max_workers = MLFLOW_TELEMETRY_MAX_WORKERS.get()
+
+        # Thread event that indicates the queue should stop processing tasks
+        self._stop_event = threading.Event()
+        self._is_active = False
+        self._atexit_callback_registered = False
+
+        self._active_tasks = set()
 
-    # temp helper function to test
-    # TODO: update this in next PR
-    def add_record(self, record: Record):
+        self._batch_size = MLFLOW_TELEMETRY_BATCH_SIZE.get()
+        self._batch_time_interval = MLFLOW_TELEMETRY_BATCH_TIME_INTERVAL.get()
+        self._pending_records: list[TelemetryRecord] = []
+        self._last_batch_time = time.time()
+        self._batch_lock = threading.Lock()
+
+    def add_record(self, record: APIRecord):
         """
-        Add a record to the queue to be sent to the telemetry server.
+        Add a record to be batched and sent to the telemetry server.
         """
-        self.records.append(record)
+        if not self.is_active():
+            self.activate()
+
+        # If stop event is set, don't add new records
+        if self._stop_event.is_set():
+            _logger.debug("Telemetry is stopped, skipping adding record")
+            return
+
+        with self._batch_lock:
+            data = self._generate_telemetry_record(record)
+            self._pending_records.append(data)
+
+            should_send = (
+                len(self._pending_records) >= self._batch_size
+                or (time.time() - self._last_batch_time) >= self._batch_time_interval
+            )
+
+            if should_send:
+                self._send_batch()
+
+    def _send_batch(self):
+        """Send the current batch of records."""
+        if not self._pending_records:
+            return
+
+        # Create a copy of the current batch and clear the pending list
+        batch_records = self._pending_records.copy()
+        self._pending_records.clear()
+        self._last_batch_time = time.time()
+
+        try:
+            self._queue.put(batch_records, block=False)
+        except Full:
+            _logger.debug("Telemetry queue is full, skipping sending data.")
+
+    def _process_records(self, records: list[TelemetryRecord]):
+        """Process a batch of telemetry records."""
+        try:
+            data = {
+                "records": [
+                    {"data": record.data, "partition-key": record.partition_key}
+                    for record in records
+                ]
+            }
+            response = requests.post(
+                self.telemetry_url,
+                json=data,
+                headers={"Content-Type": "application/json"},
+            )
+            if response.status_code != 200:
+                _logger.debug(
+                    f"Failed to send telemetry records. Status code: {response.status_code}, "
+                    f"Response: {response.text}"
+                )
+        except Exception as e:
```

Comment:
`Exception` is too broad. let's avoid using it.

Reply:
But shouldn't we capture all exceptions here?

Reply:
why?

Reply:
Yes but my concern is that if it fails with other issues, we still shouldn't block users on this?